src/edu/psu/sweng500/team8/play/GameSession.java

```diff
@@ -44,22 +45,31 @@ public Board getGameBoard() {
 	}
 	
 	/**
-	 * Undo last action such as entering a number or redo
+	 * Undo last action such as entering a number, deleting a number or redo
+	 * 
 	 */
 	public void doUndo() {
 		/* revert last action on the sudokuActionQueue, and put that action into sudokuActionQueueForUndo */
 		if (!sudokuActionQueue.isEmpty()) {
 			SudokuAction lastAction = sudokuActionQueue.pop();
+		
+			PencilMarkManager previousPencilMarkManager = lastAction.getPencilMarkManager();
+			/* TODO - we need to get the current Pencil Mark state */
+			PencilMarkManager currentPencilMarkManager = null;
+			lastAction.setPencilMarkManager(currentPencilMarkManager);
 			sudokuActionQueueForUndo.add(lastAction);
+						
+			Cell currentCell = board.getCell(lastAction.getCellCordinates().getRowIndex(), lastAction.getCellCordinates().getColumnIndex());
+			if (0 == lastAction.getPreviousValue()) {

```

Comment:
Undo/Redo is an action against the board/game and we need to keep track of a series of actions. So I put the stack in the game. 
But send me an email with what you envision on the design for redo/undo being method for action.